### PR TITLE
[2.0.0] Anime4up - Adding buttons, settings, and routing improvments

### PR DESCRIPTION
**Pull request by GhaithAlHallak8**
# With the dark logo setting
image

# Buttons setting
## With buttons
image

## Without Buttons
image

# Other
image

### CODE REVIEW

File: websites/A/Anime4up/presence.ts [hunk 1]
```diff
@@ -1,22 +1,144 @@
...
+let video = { duration: 0, currentTime: 0, paused: true };
+
+const presence = new Presence({
+    clientId: "770030754356396052"
+  }),
+  getTimestamps = (videoTime: number, videoDuration: number): number[] => {
+    const startTime = Date.now(),
+      endTime = Math.floor(startTime / 1000) - videoTime + videoDuration;
+    return [Math.floor(startTime / 1000), endTime];
+  },
```
Comment (ririxidev) [inline websites/A/Anime4up/presence.ts:38]: This function already exists in `Presence` class

File: websites/A/Anime4up/presence.ts [hunk 2]
```diff
@@ -1,22 +1,144 @@
...
+    const routes: Route[] = [
+      {
+        path: /^\/episode\//,
+        run: () => {
+          const timestamps = getTimestamps(
+            Math.floor(video.currentTime),
+            Math.floor(video.duration)
+          );
+
+          [data.startTimestamp, data.endTimestamp] = timestamps;
```
Comment (ririxidev) [inline websites/A/Anime4up/presence.ts:49]: no need for startTimestamp

File: websites/A/Anime4up/presence.ts [hunk 3]
```diff
@@ -25,82 +147,36 @@ presence.on(
...
+
+  if (showTimestamp) data.startTimestamp = Math.floor(Date.now() / 1000);
+  if (!route) return presence.setActivity(data);
+
+  if (route.run) data = route.run();
+  if (route.state) data.state = route.state();
+  if (route.details) data.details = route.details();
+  if (showButtons && route.buttons) data.buttons = route.buttons();
+  if (route.largeImageKey) data.largeImageKey = route.largeImageKey();
+  if (route.smallImageKey) data.smallImageKey = route.smallImageKey();
```
Comment (ririxidev) [inline websites/A/Anime4up/presence.ts:172]: I hope it will not be null/undefined/blank string in any way

File: websites/A/Anime4up/presence.ts [hunk 4]
```diff
@@ -25,82 +147,36 @@ presence.on(
...
+  if (route.smallImageText) data.smallImageText = route.smallImageText();
+
+  if (showTimestamp && route.startTimestamp)
+    data.startTimestamp = route.startTimestamp();
```
Comment (ririxidev) [inline websites/A/Anime4up/presence.ts:176]: probs you can remove it like I said above

**Review by rxri — CHANGES_REQUESTED** [review]

File: websites/A/Anime4up/presence.ts (hunk 1 above)
Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:38]: nice. didn't know about it

File: websites/A/Anime4up/presence.ts (hunk 3 above)
Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:172]: no, it's safe

File: websites/A/Anime4up/presence.ts (hunk 2 above)
Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts]: yea sorry i forget to remove it

**Review by GhaithAlHallak8 — COMMENTED** [review]
oh sorry #bba12feefff125c004a18439b08f44c8f66b8da1 shouldn't be here, forget to create another branch

**Review by Bas950 — DISMISSED** [review]

File: websites/A/Anime4up/presence.ts [hunk 5]
```diff
@@ -1,22 +1,161 @@
...
+          `Episode: ${document
+            .querySelectorAll(".container")[1]
+            .textContent.slice(1, -1)
+            .split(" ")
+            .pop()}`,
+
+        buttons: () => [
+          { label: "Watch Episode", url: location.href },
+          {
+            label: "Anime Page",
```
Comment (ririxidev) [inline websites/A/Anime4up/presence.ts:81]:
```suggestion
            label: "View Page",
```

File: websites/A/Anime4up/presence.ts [hunk 6]
```diff
@@ -1,22 +1,161 @@
...
+
+        details: () => "Viewing an Anime",
+
+        state: () => document.querySelector(".anime-details-title").textContent,
+
+        buttons: () => [
+          { label: "View Anime", url: location.href },
+          {
+            label: "Last Episode",
+            // prettier-ignore
```
Comment (ririxidev) [inline websites/A/Anime4up/presence.ts:123]: no prettier ignore 😠

**Review by rxri — CHANGES_REQUESTED** [review]
prettier plz, there are new lines for no reason :/

File: websites/A/Anime4up/presence.ts (hunk 5 above)
Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:81]: there's a misunderstanding here, the viewed page will be the episode's page, not the anime's, so Anime Page is more accurate

Watch Episode => current location
Anime Page => the current viewed episode's anime location

Comment (ririxidev) [reply websites/A/Anime4up/presence.ts:81]: There is no misunderstanding. Then please change it to something else than `Anime Page`

File: websites/A/Anime4up/presence.ts (hunk 6 above)
Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:123]: unfortunately, I've to put it, for some reason when I format the file via prettier, eslint will suddenly throw an error

Comment (ririxidev) [reply websites/A/Anime4up/presence.ts:123]: What eslint throws then?

Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:123]: image

Unexpected newline between object and [ of property access.eslintno-unexpected-multiline

Comment (ririxidev) [reply websites/A/Anime4up/presence.ts:123]: Then try assigning this value to variable bc it looks like parser issue from eslint

Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:123]: and the same thing for GMANGA presence right?

Comment (ririxidev) [reply websites/A/Anime4up/presence.ts:123]: This is in GMANGA too? Well yeah then

File: websites/A/Anime4up/presence.ts (hunk 5 above)
Comment (GhaithAlHallak8) [reply websites/A/Anime4up/presence.ts:81]: then View Anime?

Comment (ririxidev) [reply websites/A/Anime4up/presence.ts:81]: Yeah, better.

**Review by Bas950 — DISMISSED** [review]

**Comment by ririxidev** [comment]
still prettify, thanks (remove blank lines)

**Comment by Bas950** [comment]
> still prettify, thanks (remove blank lines)

doesn't SeMiD do that tho?

**Comment by ririxidev** [comment]
SeMiD will not remove these files, and it's always to have good style at start than giving everything to semid :/

**Comment by GhaithAlHallak8** [comment]
> 
> 
> still prettify, thanks (remove blank lines)

yea? I format it already, I don't see any blank lines

**Comment by ririxidev** [comment]
image
I do

**Comment by GhaithAlHallak8** [comment]
> 
> 
> image
> I do

so you mean those, I thought they are more readable & beautiful

**Comment by GhaithAlHallak8** [comment]
> 
> 
> > image
> > I do
> 
> so you mean those, I thought they are more readable & beautiful

btw np

**Comment by ririxidev** [comment]
Problem is it doesn't look good and just makes code longer 😂

**Review by rxri — APPROVED** [review]

**Review by Slowlife01 — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**